docs(parachain): design for prospective-parachains (#4348)
ChainSafe/gossamer: docs/docs/design/parachain-protocol/fragment-chain.md

# Fragment Chain 

### Candidate Storage

> This structure does not care if candidates form a chain, it only stores candidates and holds "links" to a `CandidateEntry`, those links are

Reviewer: Can you add links to the polkadot repo code for these types?

`Parent Head Data Hash` to `Set<Candidate Hash>`
`Output Head Data Hash` to `Set<Candidate Hash>`

The parent head data hash is simply the parent block of that candidate, the output head data hash is the hash of the resultant `HeadData` after an execution, you can think of the output head data as the raw head of the parachain block that we get after validating the inputs against the parachain runtime.

> `input -> execution -> head data (parachain block head)`, it is called head data because it is a `Vec<u8>`, what this head data contains is not for the validator, what we care is the hash of it!

Reviewer: ```suggestion
`input -> execution -> head data (parachain block head)`, it is called head data because it is a `Vec<u8>`, what this head data contains is not for the validator, what we care is the *hash* of it.
```

### Scope

Scope is the context of a fragment chain, it contains constraints and ancestors that rules candidates, e.g. if a candidate was build in the context of an allowed relay parent.

### Fragment Node

> This is a node that is part of a `BackedChain`. When we create a new fragment we should make sure that the candidate is valid under the operating constraints. We use the constraints to build an expected persisted validation data and compare with the one we got from the candidate. A fragment node is always backed and it is generated from a `Candidate Entry`

Reviewer: ```suggestion
This is a node that is part of a `BackedChain`. When we create a new fragment we should make sure that the candidate is valid under the operating constraints. We use the constraints to build an expected persisted validation data and compare with the one we got from the candidate. A fragment node is always backed and it is generated from a `CandidateEntry`.
```

### BackedChain

It holds a chain of backed nodes (fragments), in rust they store the fragments.

> ###### RevertToParentHash

Reviewer: ```suggestion
#### RevertToParentHash
```

```
{Parent Head Hash (0) | Output Head Data Hash (1)}
|
V
{Parent Head Hash (1) | Output Head Data Hash (2)}
|
V
{Parent Head Hash (2) | Output Head Data Hash (3)}
|
V
{Parent Head Hash (4) | Output Head Data Hash (5)}
|
V
{Parent Head Hash (5) | Output Head Data Hash (6)}
```

If we want to revert to `ParentHeadDataHash (3)` we search for a node that outputs such head data hash, in this case is the node at position 2, knowing the index we should remove its children (nodes at position 3 and 4) and clear their informations from the tables `byParentHead`, `byOutputHead` and `Candidates`, and we return the removed nodes.

### Fragment Chain

The fragment chain is a struct that an active leaf holds, also it should avoid cycles or paths in the three.

##### Populate Chain

This is a method from `FragmentChain` where given a scope and a Candidate Storage it will try to build a `BackedChain` with the candidates that follow the conditions:

- Does not introduce a cycle: That means, its output head data hash is not the parent head data hash of another candidate in the BackedChain.
- Its parent hash is correct: matches the previous output hash, forming a coherent chain

> - Relay parent does not move backwards: given our earliest relay parent (the earliest ancestor that a candidate can use as its context) the current candidate's relay parent should not be befor it.

Reviewer: ```suggestion
- Relay parent does not move backwards: given our earliest relay parent (the earliest ancestor that a candidate can use as its context) the current candidate's relay parent should not be before it.
```

- All non-pending-availability has a parent in the current scope.
- Candidates outputs fullfils the constraint.

If those conditions follows for 1 or more candidates we consider them possible backed/backable, however since the `BackedChain` does not allow forks we need to resolve that if there is more than one candidates, and the fork selection rule is: compare the hashes and the lowest hash will be chosen and pushed into the `BackedChain` and removed from the Candidate Storage

##### Trim Unelegible Forks

What is does is a deapth breadth-first search looking for candidates in the Candidate Storage that has no potential that is children of valid backed candidates.

Consider the following chain, the green letters forms the `BackedChain` while the red letters are candidates placed in the `Unconnected` candidate storage, drawing them (without all the maps and hashes here and there) we can have something similar to the image bellow.

> ![IMG_944F8C1D62F0-1](https://hackmd.io/_uploads/SJ3KA0m7kg.jpg)

Reviewer 2: I do not have access to this image and overall, I guess, it would be better if we upload images to repo. wdyt?

Now lets trim the unelegible forks out, but first we need to define the steps:

> 1. Lets build a queue of candidates in the backed chain (a.k.a green letters) and we will use an extra info called has potential, saying that the candidate is good to stay connected (even if they form a fork).

Reviewer: ```suggestion
1. Lets build a queue of candidates in the backed chain (a.k.a green letters) and we will use an extra info called "has potential", which means that the candidate is good to stay connected (even if they form a fork).
```

2. queue is empty, the goto step 5, queue is not empty, pop the item from the start of the queue, mark it as visited and iterate over its children.

> 3. for each children:

Reviewer: ```suggestion
3. for each child:
```

>     3.1. is the children in the unconnected hash?

Reviewer: ```suggestion
    3.1. is the child in the unconnected hash set?
```

        - **true**:  go to next step
        - **false**: go to step 1

>     3.2. did we already visited this children?

Reviewer: ```suggestion
    3.2. did we already visited this child?
```

        - **true**: go to step 1
        - **false**: go to next step

>     3.3. is the candidate and the child potentials?

Reviewer: ```suggestion
    3.3. does the candidate and the child have potential?
```

        - **true**:  add to the queue as (child hash, true)
        - **false**: mark to remove, add to the queue as (child hash, false)

4. go to step 1

5. remove all items marked to be removed.

> Lets apply the chain from image against the steps defined above, first things first, we should build the queue from our `backed chain`.

Reviewer: ```suggestion
Lets apply the chain from image against the steps defined above, first things first, we should build the queue from our backed chain.
```

```
QUEUE     -> [ (A, true), (B, true), (C, true), (D, true) ]
VISITED   -> []
TO_REMOVE -> 
```

jumping to the next instruction, lets pop `(A, true)` from the queue, mark it as visited and iterate over its children!

```
QUEUE     -> [ (B, true), (C, true), (D, true) ]
VISITED   -> [ A ]
TO_REMOVE -> []
```

> ![IMG_88FB30D4B931-1](https://hackmd.io/_uploads/rysElgE71x.jpg)

Reviewer 2: same as above about images

them, child `B1` is in the unconnected set, is not visited, its parent (`A`) is potential, however `B1` is not, them we should mark it to remove and add to the queue as (`B1`, `false`) 

```
QUEUE     -> [ (B, true), (C, true), (D, true), (B1, false) ]
VISITED   -> [ A ]
TO_REMOVE -> [ B1 ]
```

the next child is `B`, which is not in the unconnected set, them since there is no children left we can just jump to the step 1.

continuing, lets pop `(B, true)` from the queue, mark it as visited and iterate over its children!

```
QUEUE     -> [ (C, true), (D, true), (B1, false) ]
VISITED   -> [ A, B ]
TO_REMOVE -> [ B1 ]
```

![IMG_319583A1BB18-1](https://hackmd.io/_uploads/H1f7GeE71e.jpg)

the node `B`, has children `C` and `C1`, starting with `C`, given `C` is in the `backed chain` it is not in the unconnected set, in this case we can go to the next child that is `C2`, that is in the unconnected set, was not visited, and has a parent (`B`) that is potential but `C2` itself is not, then lets push it to the queue as `(C2, false)` and mark it to remove.

```
QUEUE     -> [ (C, true), (D, true), (B1, false), (C2, false) ]
VISITED   -> [ A, B ]
TO_REMOVE -> [ B1, C2 ]
```

> as there is no more children... lets pop `(C, true)` from the queue, mark it as visited and iterate over its children!

Reviewer: ```suggestion
as there are no more children... lets pop `(C, true)` from the queue, mark it as visited and iterate over its children!
```
